base-permissions/src/commonMain/kotlin/Permission.kt

```diff
@@ -0,0 +1,133 @@
+/*
+ Copyright (c) 2020. Splendo Consulting B.V. The Netherlands
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+ */
+
+package com.splendo.kaluga.permissions
+
+import co.touchlab.stately.collections.IsoMutableMap
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.transformLatest
+import kotlin.coroutines.CoroutineContext
+import kotlin.reflect.KClassifier
+
+/**
+ * Permissions that can be requested by Kaluga
+ * Each permission component should declare subclass of [Permission]
+ */
+abstract class Permission
+
+/**
+ *  Base type for all permissions builders. Each permission component implements a concrete permissions builder
+ *  and declares helper methods for registering this builder in the [PermissionsBuilder].
+ */
+interface BasePermissionsBuilder
+
+/**
+ * Closure that takes permission and coroutine contexts and creates [PermissionStateRepo].
+ * Each platform registers [RepoFactory] in the register permission helper method.
+ */
+typealias RepoFactory = (permission: Permission, coroutineContext: CoroutineContext) -> PermissionStateRepo<*>
+
+/**
+ * Builder for providing the proper [PermissionManager] for each [Permission]
+ * @param context an additional parameter platform can pass to the [PermissionsBuilder]. [NSBundle] on iOS and [Contect] on Andoid.
+ * To create builder with parameter on a platform an extension of the companion object [Factory] can be used.
+ * it will be [PermissionsBuilder.withBundle] on iOS and [PermissionsBuilder.withContext] on Android.
+ */
+open class PermissionsBuilder(val context: Any? = null) {
```

Comment:
From the looks of it, every single permission has the same context type. And for each you need to do the same instance check.

Why not just create expect class PermissionContext and declare a type alias for iOS and Android